InstCombine: Order shufflevector operands by complexity

MatzeB opened the pull request:
As shufflevector is effectively commutative we should apply the same
logic as other commutative operations where we order the inputs by
their `getComplexity()` value. This will put things like `undef`,
`poison` and constants on the right hand side where possible.

We had a rule moving `undef` to the right hand side that is superseded
by this.

MatzeB committed:
InstCombine: Order shufflevector operands by complexity

As shufflevector is effectively commutative we should apply the same
logic as other commutative operations where we order the inputs by
their `getComplexity()` value. This will put things like `undef`,
`poison` and constants on the right hand side where possible.

We had a rule moving `undef` to the right hand side that is superseded
by this.

efriedma-quic commented:
Please push the revised clang tests as a separate patch.

This idea here makes sense, I guess, but I'm a little worried about unexpected side-effects... not sure how reliably the backend handles commuted masks, and I don't think we have any testing infrastructure that would catch that sort of issue.

goldsteinn commented:
@dtcxzyw do you still support riscv codegen diffs?

MatzeB committed:
Remove -O3 in clang codegen test

- Remove an -O3 flag from a couple of clang x86 codegen tests so the
  tests do not need to be updated when optimizations in LLVM change.
- Change the tests to use utils/update_cc_test_checks.sh
- Change from apple/darwin triples to generic x86_64-- and
  i386-- because it was not relevant to the test but
  `update_cc_test_checks` seems to be unable to handle platforms that
  prepend `_` to function names.

MatzeB commented:
> This idea here makes sense, I guess, but I'm a little worried about unexpected side-effects... not sure how reliably the backend handles commuted masks, and I don't think we have any testing infrastructure that would catch that sort of issue.

Was wondering about this too, but at least it should be easy to fix things if some targets somehow depend on the order and produce worse code in one ordering but not the other. Given that these targets could have reached this situation without this change it may be a good thing to flush out this behavior...

MatzeB commented on clang/test/CodeGen/X86/avx-shuffle-builtins.c:
```diff
@@ -1,7 +1,7 @@
-// RUN: %clang_cc1 -ffreestanding %s -O3 -triple=i386-apple-darwin -target-feature +avx -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK,X86
-// FIXME: This is testing optimized generation of shuffle instructions and should be fixed.
+// RUN: %clang_cc1 -ffreestanding %s -target-feature +avx -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK
+// RUN: %clang_cc1 -ffreestanding %s -target-feature +avx -emit-llvm -o - | FileCheck %s --check-prefixes=CHECK
```
My bad, I wanted to drop the "apple-darwin" part because apparently `update_cc_test_checks.sh` cannot deal with the `_` prefix added on apple/darwin platforms. Made sure I keep an x64_64 and i386 triple around in the new updates in #113714

MatzeB commented:
> Please push the revised clang tests as a separate patch.

Submitted #113714 for the test updates (still included here for "stacking")

davemgreen reviewed (commented):
I have always expected shuffles to be canonicalized to make the lowest mask lane the first operand. I believe the AArch64 and Arm matching functions rely on that at the moment. https://godbolt.org/z/1rr1E8v1K

MatzeB commented:
> I have always expected shuffles to be canonicalized to make the lowest mask lane the first operand. I believe the AArch64 and Arm matching functions rely on that at the moment. https://godbolt.org/z/1rr1E8v1K

Thanks for pointing this out! So definitely cannot land the change as-is. I would still argue it is generally the right thing to do: We definitely need to normalize by some rule to enable CSE and hit better code generation path (as your example also demonstrates).

Normalizing for low complexity on RHS comes from working on #113746 which slightly simplifies things when we can assume constants being always on the RHS. And this also being a well established pattern for arithmetic operations (add, mul, bitops). But I think that doesn't rule out normalizing for something different at ISel time.

Let me investigate the situation for the popular targets. FWIW: I don't strictly need this rule for my immediate work; this is more of a drive-by fix that seemed worth doing to me. So happy to drop this change if people aren't comfortable with it...

davemgreen commented:
It sounds OK so lang as we can make sure the backend patterns keep working - it sounds like it would be more resilient overall if we matched both forms. I think it was just assumed in the past that is wasn't needed.

MatzeB commented:
I did some experiments with randomly commuting `shufflevector`s before instruction selection and also experimented with additional canonicalization in SelectionDAG DAGCombine phase. All in all this produces quite a bit of churn and noise for X86 and AArch64 and I feel somewhat uncomfortable pushing this now. Fixing the backends to be resilient and consistently produce the best code seems like a bigger project, so I think I will not move forward with this PR for now.

Here's my experiments: https://github.com/MatzeB/llvm-project/commits/shuffle_rando_experiment/  and especially the last commit adjusting the test-cases shows that the backend patterns are currently fragile...
